receiver/awsfirehosereceiver/internal/unmarshaler/otlpmetricstream/unmarshaler.go

```diff
@@ -0,0 +1,98 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package otlpmetricstream // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/internal/unmarshaler/otlpmetricstream"
+
+import (
+	"errors"
+
+	"github.com/gogo/protobuf/proto"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/internal/unmarshaler"
+)
+
+const (
+	// Supported version depends on version of go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp dependency
+	TypeStr = "otlp_v1"
+)
+
+var (
+	errInvalidOTLPFormatStart = errors.New("unable to decode data length from message")
+)
+
+// Unmarshaler for the CloudWatch Metric Stream OpenTelemetry record format.
+//
+// More details can be found at:
+// https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats-opentelemetry-100.html
+type Unmarshaler struct {
+	logger *zap.Logger
+}
+
+var _ unmarshaler.MetricsUnmarshaler = (*Unmarshaler)(nil)
+
+// NewUnmarshaler creates a new instance of the Unmarshaler.
+func NewUnmarshaler(logger *zap.Logger) *Unmarshaler {
+	return &Unmarshaler{logger}
+}
+
+// Unmarshal deserializes the records into pmetric.Metrics
+func (u Unmarshaler) Unmarshal(records [][]byte) (pmetric.Metrics, error) {
+	md := pmetric.NewMetrics()
+	for recordIndex, record := range records {
+		var dataLen, pos = len(record), 0
+		for pos < dataLen {
+			n, nLen := proto.DecodeVarint(record)
+			if nLen == 0 && n == 0 {
+				return md, errInvalidOTLPFormatStart
+			}
+			req := pmetricotlp.NewExportRequest()
+			pos += nLen
+			err := req.UnmarshalProto(record[pos : pos+int(n)])
+			pos += int(n)
+			if err != nil {
+				u.logger.Error(
+					"Unable to unmarshal input",
+					zap.Error(err),
+					zap.Int("record_index", recordIndex),
+				)
+				continue
+			}
+			if !u.isValid(req.Metrics()) {
+				u.logger.Error(
+					"Invalid metric",
+					zap.Int("record_index", recordIndex),
+				)
+				continue
+			}
+			req.Metrics().ResourceMetrics().MoveAndAppendTo(md.ResourceMetrics())
+		}
+	}
+
+	return md, nil
+}
+
+// isValid validates that the metric has been unmarshalled correctly.
+func (u Unmarshaler) isValid(metrics pmetric.Metrics) bool {
+	for r := range metrics.ResourceMetrics().Len() {
+		for s := range metrics.ResourceMetrics().At(r).ScopeMetrics().Len() {
+			for m := range metrics.ResourceMetrics().At(r).ScopeMetrics().At(s).Metrics().Len() {
+				if metrics.ResourceMetrics().At(r).ScopeMetrics().At(s).Metrics().At(m).Name() == "" {
+					return false
+				}
+				if metrics.ResourceMetrics().At(r).ScopeMetrics().At(s).Metrics().At(m).Type() == pmetric.MetricTypeEmpty {
+					return false
```

Comment:
Does it make sense to do this validation here?  This requires evaluating every metric received to be evaluated for things that can't be represented in OTLP (`pmetric.MetricTypeEmpty`).  The `otlpreceiver` doesn't even bother with this.  If it can be unmarshaled from the protobuf without error then it is well-formed and can be passed on to the next consumer.

Reply:
Hi, our colleague working on this is on vacation and we will get back to this next week. Just FYI.

Reply:
Removed extra validation as it already checks if unmarshal is successful.